### ISSUE TITLE
segmentation violation on interrupt

### ISSUE THREAD

**Issue by businessfriendlyusername** [ethereum/go-ethereum#20585]
#### System information

Geth version: `1.9.10-stable`
OS & Version: Ubuntu 18.04.3
Commit hash : 58cf5686eab9019cc01e202e846a6bbc70a3301d

#### Expected behaviour
expected graceful shutdown on keyboard interrupt

#### Actual behaviour

segmentation violation
```
INFO [01-21|01:07:26.843] Got interrupt, shutting down... 
INFO [01-21|01:07:27.044] Imported new state entries               count=366  elapsed=1.227ms      processed=63737646 pending=8865  retry=0   duplicate=139 unexpected=1055
INFO [01-21|01:07:27.058] IPC endpoint closed                      url=/home/USER/.ethereum/geth.ipc
INFO [01-21|01:07:27.249] Blockchain manager stopped 
INFO [01-21|01:07:27.289] Stopping Ethereum protocol 
INFO [01-21|01:07:27.332] Ethereum protocol stopped 
INFO [01-21|01:07:27.332] Transaction pool stopped 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x957723]
```
#### Steps to reproduce the behaviour
not reproducible to my knowledge

#### Backtrace

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x957723]

goroutine 32039 [running]:
github.com/ethereum/go-ethereum/core/types.(*Block).NumberU64(...)
	/build/ethereum-k5K5KF/ethereum-1.9.10+build21009+bionic/core/types/block.go:309
github.com/ethereum/go-ethereum/core.(*BlockChain).Rollback(0xc004dae000, 0xc02fd29800, 0x12b, 0x12b)
	/build/ethereum-k5K5KF/ethereum-1.9.10+build21009+bionic/core/blockchain.go:921 +0x5a3
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).processHeaders.func1(0xc024639d90, 0xc004fe0000)
	/build/ethereum-k5K5KF/ethereum-1.9.10+build21009+bionic/eth/downloader/downloader.go:1374 +0x1e5
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).processHeaders(0xc004fe0000, 0x8e45a8, 0x8e4441, 0xc0ef3feec0, 0x13d6180, 0xc0001d2490)
	/build/ethereum-k5K5KF/ethereum-1.9.10+build21009+bionic/eth/downloader/downloader.go:1393 +0x12cc
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).syncWithPeer.func6(0xc0137a4f70, 0x671efb301529ee16)
	/build/ethereum-k5K5KF/ethereum-1.9.10+build21009+bionic/eth/downloader/downloader.go:517 +0x48
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).spawnSync.func1(0xc004fe0000, 0xc01724b200, 0xc0130e05d0)
	/build/ethereum-k5K5KF/ethereum-1.9.10+build21009+bionic/eth/downloader/downloader.go:534 +0x63
created by github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).spawnSync
	/build/ethereum-k5K5KF/ethereum-1.9.10+build21009+bionic/eth/downloader/downloader.go:534 +0xaf

```

**Comment by timkuijsten** [comment]
I can confirm the same issue with geth 1.9.10-stable on macOS 10.15.2.

**Comment by PyrateAkananto** [comment]
I can confirm the same issue with geth 1.9.10 on Windows 10 (`geth-windows-amd64-1.9.10-58cf5686.zip`). Happens every time I press Ctrl-C within the `cmd.exe` window. `geth.exe` was started with `--syncmode fast`. Never fully syncs, always around 100 blocks missing.

**Comment by PyrateAkananto** [comment]
Still happens with geth 1.9.11 on Windows 10 (`geth-windows-amd64-1.9.11-6a62fe39.zip`).

```
INFO [03-04|10:27:12.082] Got interrupt, shutting down...
INFO [03-04|10:27:12.602] HTTP endpoint closed                     url=http://127.0.0.1:8545
INFO [03-04|10:27:12.605] IPC endpoint closed                      url=\\\\.\\pipe\\geth.ipc
INFO [03-04|10:27:12.608] Blockchain manager stopped
INFO [03-04|10:27:12.668] Stopping Ethereum protocol
INFO [03-04|10:27:12.672] Ethereum protocol stopped
INFO [03-04|10:27:12.673] Transaction pool stopped
WARN [03-04|10:27:12.876] Synchronisation failed, retrying         err="DB write error: leveldb: closed"
INFO [03-04|10:27:12.894] Initialized fast sync bloom              items=6462708 errorrate=0.000 elapsed=3m38.684s
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x1c0 pc=0x95fe19]

goroutine 494 [running]:
github.com/ethereum/go-ethereum/core.(*HeaderChain).SetCurrentHeader(0xc0000c4360, 0x0)
        C:/gopath/src/github.com/ethereum/go-ethereum/core/headerchain.go:477 +0xa9
github.com/ethereum/go-ethereum/core.(*BlockChain).Rollback(0xc000e86000, 0xc089cd0000, 0x800, 0x800)
        C:/gopath/src/github.com/ethereum/go-ethereum/core/blockchain.go:915 +0x380
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).processHeaders.func1(0xc07128bd90, 0xc00061e000)
        C:/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:1374 +0x1ec
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).processHeaders(0xc00061e000, 0x71b41f, 0x71b3d7, 0xc003eea240, 0x16bd540, 0xc0000f1c00)
        C:/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:1393 +0x12d3
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).syncWithPeer.func6(0xc04c41ff70, 0xa932c7)
        C:/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:517 +0x4f
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).spawnSync.func1(0xc00061e000, 0xc006998f00, 0xc002dbc5a0)
        C:/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:534 +0x6a
created by github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).spawnSync
        C:/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:534 +0xb6
Batchvorgang abbrechen (J/N)?
```

**Comment by ligi** [comment]
This should not happen anymore - if it does again please reopen.
